application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramContentHandler.java

```diff
@@ -174,6 +174,8 @@ public void updateDiagramContent(XWikiDocument backlinkDoc, DocumentReference or
     public List<DocumentReference> getLinkedPages(String content, DocumentReference diagramReference)
     {
         try {
+            getDiagramLinksHandler.resetLinkedPages();
```

Comment:
This made me realize that you have an issue: ``GetDiagramLinksHandler`` is marked as **singleton** (shared by all threads / requests) but it has a **state** (``linkedPages``). This is a recipe for disaster unless you make the state **thread-safe** but this also makes it slower. Best is to:

* replace the ``@Singleton`` annotation from ``GetDiagramLinksHandler`` with ``@InstantiationStrategy(ComponentInstantiationStrategy.PER_LOOKUP)``
* use ``Provider<GetDiagramLinksHandler>`` in ``DiagramContentHandler`` in order to make sure a new instance is created for each ``getLinkedPages`` call -> this removes the need for a ``resetLinkedPages()`` method.